contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts

```diff
@@ -0,0 +1,851 @@
+
+blockchainTests('erc20-bridge-sampler', env => {
+    let testContract: TestERC20BridgeSamplerContract;
+    let allSources: { [name: string]: string };
+    const RATE_DENOMINATOR = constants.ONE_ETHER;
+    const MIN_RATE = new BigNumber('0.01');
+    const MAX_RATE = new BigNumber('100');
+    const MIN_DECIMALS = 4;
+    const MAX_DECIMALS = 20;
+    const WETH_ADDRESS = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2';
+    const KYBER_SALT = '0x0ff3ca9d46195c39f9a12afb74207b4970349fb3cfb1e459bbf170298d326bc7';
+    const ETH2DAI_SALT = '0xb713b61bb9bb2958a0f5d1534b21e94fc68c4c0c034b0902ed844f2f6cd1b4f7';
+    const UNISWAP_BASE_SALT = '0x1d6a6a0506b0b4a554b907a4c29d9f4674e461989d9c1921feb17b26716385ab';
+    const ERC20_PROXY_ID = '0xf47261b0';
+    const INVALID_ASSET_PROXY_ASSET_DATA = hexConcat('0xf47261b1', hexLeftPad(randomAddress()));
+    const INVALID_ASSET_DATA = hexRandom(37);
+    const SELL_SOURCES = ['Eth2Dai', 'Kyber', 'Uniswap'];
+    const BUY_SOURCES = ['Eth2Dai', 'Uniswap'];
+    const EMPTY_ORDERS_ERROR = 'EMPTY_ORDERS';
+    const UNSUPPORTED_ASSET_PROXY_ERROR = 'UNSUPPORTED_ASSET_PROXY';
+    const INVALID_ASSET_DATA_ERROR = 'INVALID_ASSET_DATA';
+    const UNSUPPORTED_UNISWAP_EXCHANGE_ERROR = 'UNSUPPORTED_UNISWAP_EXCHANGE';
+    const UNSUPPORTED_SOURCE_ERROR = 'UNSUPPORTED_SOURCE';
+    const INVALID_TOKEN_PAIR_ERROR = 'INVALID_TOKEN_PAIR';
+
+    before(async () => {
+        testContract = await TestERC20BridgeSamplerContract.deployFrom0xArtifactAsync(
+            artifacts.TestERC20BridgeSampler,
+            env.provider,
+            env.txDefaults,
+            {},
+        );
+        allSources = _.zipObject(
+            ['Uniswap', 'Eth2Dai', 'Kyber'],
+            [
+                await testContract.uniswap().callAsync(),
+                await testContract.eth2Dai().callAsync(),
+                await testContract.kyber().callAsync(),
+            ],
+        );
+    });
+
+    function getPackedHash(...args: string[]): string {
+        return hexHash(hexConcat(...args.map(a => toHex(a))));
+    }
+
+    function getUniswapExchangeSalt(tokenAddress: string): string {
+        return getPackedHash(UNISWAP_BASE_SALT, tokenAddress);
+    }
+
+    function getDeterministicRate(salt: string, sellToken: string, buyToken: string): BigNumber {
+        const hash = getPackedHash(salt, sellToken, buyToken);
+        const _minRate = RATE_DENOMINATOR.times(MIN_RATE);
+        const _maxRate = RATE_DENOMINATOR.times(MAX_RATE);
+        return new BigNumber(hash)
+            .mod(_maxRate.minus(_minRate))
+            .plus(_minRate)
+            .div(RATE_DENOMINATOR);
+    }
+
+    function getDeterministicTokenDecimals(token: string): number {
+        if (token === WETH_ADDRESS) {
+            return 18;
+        }
+        return (
+            (new BigNumber(getPackedHash(token)).mod(MAX_DECIMALS - MIN_DECIMALS).toNumber() as number) + MIN_DECIMALS // linter is confused
```

Comment:
wot

Reply:
For some reason `toNumber()` wasn't being picked up as a `number`.

Reply:
I'll clean it up in the rebase.